In [5]:
import pandas as pd
import matplotlib.pyplot as plt

In [6]:
df = pd.read_csv("../data/StudentsPerformance.csv")

# Dataset Overview
print(df.head(3))
print(df.shape)
print(df.columns)
print(df.info())

   gender race/ethnicity parental level of education     lunch  \
0  female        group B           bachelor's degree  standard   
1  female        group C                some college  standard   
2  female        group B             master's degree  standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
(1000, 8)
Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 

## Observations

- The dataset contains 1000 student records and 8 columns.
- There are 5 categorical features and 3 numerical features.
- The numerical columns represent scores in Mathematics, Reading, and Writing.
- No missing values were found in any column.
- The dataset appears clean and suitable for further analysis.
- Features such as gender, parental education, lunch type, and test preparation course may help identify factors affecting student performance.

In [7]:
# Data cleaning
df.isnull().sum()
df.duplicated()
df.drop_duplicates()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75
...,...,...,...,...,...,...,...,...
995,female,group E,master's degree,standard,completed,88,99,95
996,male,group C,high school,free/reduced,none,62,55,55
997,female,group C,high school,free/reduced,completed,59,71,65
998,female,group D,some college,standard,completed,68,78,77


## Observations

- No missing values were found in any column.
- No duplicate records were detected in the dataset.
- The dataset appears clean and does not require major preprocessing.
- The data is ready for exploratory analysis.

In [8]:
# Statistical Analysis

# Summary Statistics
df.describe()


,math score,reading score,writing score
count,1000.00000,1000.000000,1000.000000
mean,66.08900,69.169000,68.054000
std,15.16308,14.600192,15.195657
min,0.00000,17.000000,10.000000
25%,57.00000,59.000000,57.750000
50%,66.00000,70.000000,69.000000
75%,77.00000,79.000000,79.000000
max,100.00000,100.000000,100.000000


In [9]:
# Average Score of Each Subject

subject_means = df[["math score", "reading score", "writing score"]].mean()
print(subject_means)

math score       66.089
reading score    69.169
writing score    68.054
dtype: float64


In [10]:
# Highest Average Subject
highest_avg_subject = subject_means.idxmax()
print("Highest Average Subject:", highest_avg_subject)

# Lowest Average Subject
lowest_avg_subject = subject_means.idxmin()
print("Lowest Average Subject:", lowest_avg_subject)

Highest Average Subject: reading score
Lowest Average Subject: math score


In [11]:
# Highest Score in Each Subject
print("Highest Score in Each Sub:")
print(df[["math score", "reading score", "writing score"]].max())

# Lowest Score in Each Subject
print("\nLowest Score in Each Sub:")
print(df[["math score", "reading score", "writing score"]].min())

Highest Score in Each Sub:
math score       100
reading score    100
writing score    100
dtype: int64

Lowest Score in Each Sub:
math score        0
reading score    17
writing score    10
dtype: int64


## Observations

- The average Reading score (69.17) is the highest among all subjects.
- The average Mathematics score (66.09) is the lowest among all subjects.
- Students generally perform better in Reading and Writing compared to Mathematics.
- Perfect scores (100) were achieved in all three subjects, indicating the presence of high-performing students.
- The minimum Mathematics score is 0, while the minimum Reading and Writing scores are 17 and 10 respectively, showing significant variation in student performance.

In [12]:
df["average score"] = (
    df["math score"] + 
    df["reading score"]+
    df["writing score"]
) / 3

df

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score,average score
0,female,group B,bachelor's degree,standard,none,72,72,74,72.666667
1,female,group C,some college,standard,completed,69,90,88,82.333333
2,female,group B,master's degree,standard,none,90,95,93,92.666667
3,male,group A,associate's degree,free/reduced,none,47,57,44,49.333333
4,male,group C,some college,standard,none,76,78,75,76.333333
...,...,...,...,...,...,...,...,...,...
995,female,group E,master's degree,standard,completed,88,99,95,94.000000
996,male,group C,high school,free/reduced,none,62,55,55,57.333333
997,female,group C,high school,free/reduced,completed,59,71,65,65.000000
998,female,group D,some college,standard,completed,68,78,77,74.333333


In [13]:
# Performance Analysis
df2 = df[["gender", "math score", "reading score", "writing score", "average score"]]

# Top 10 students
top_ten = df2.sort_values("average score", ascending=False).head(10)


# Bottom 10 students
bottom_ten = df2.sort_values("average score").head(10)

In [14]:
top_ten

,gender,math score,reading score,writing score,average score
916,male,100,100,100,100.000000
962,female,100,100,100,100.000000
458,female,100,100,100,100.000000
114,female,99,100,100,99.666667
712,female,98,100,99,99.000000
179,female,97,100,100,99.000000
165,female,96,100,100,98.666667
625,male,100,97,99,98.666667
685,female,94,99,100,97.666667
903,female,93,100,100,97.666667


In [15]:
bottom_ten

,gender,math score,reading score,writing score,average score
59,female,0,17,10,9.000000
980,female,8,24,23,18.333333
596,male,30,24,15,23.000000
327,male,28,23,19,23.333333
17,female,18,32,28,26.000000
76,male,30,26,22,26.000000
601,female,29,29,30,29.333333
338,female,24,38,27,29.666667
787,female,19,38,32,29.666667
211,male,35,28,27,30.000000


## Observations

- The top-performing students consistently achieved high scores across all three subjects.

- Several top-performing students obtained scores close to or equal to 100 in multiple subjects.

- The bottom-performing students showed low performance across multiple subjects rather than a single subject.

- Average score provides a balanced measure for evaluating overall student performance.

In [16]:
# Gender-Based Analysis

df.groupby("gender")[
    ["math score", 
     "reading score", 
     "writing score", 
     "average score"
    ]
].mean()

,math score,reading score,writing score,average score
gender,,,,
female,63.633205,72.608108,72.467181,69.569498
male,68.728216,65.473029,63.311203,65.837483


## Observations

* Female students achieved a higher overall average score (69.57) compared to male students (65.84).

* Female students performed significantly better in Reading and Writing, with average scores above 72 in both subjects.

* Male students performed better in Mathematics, achieving an average score of 68.73 compared to 63.63 for female students.

* Despite stronger Mathematics performance among male students, female students demonstrated better overall academic performance due to higher Reading and Writing scores.

* The results suggest that performance differences exist across subjects between male and female students.

In [17]:
# Test Preparation Analysis

df.groupby("test preparation course") [
    ["math score", 
     "reading score", 
     "writing score", 
     "average score"
    ]
].mean()

,math score,reading score,writing score,average score
test preparation course,,,,
completed,69.695531,73.893855,74.418994,72.669460
none,64.077882,66.534268,64.504673,65.038941


## Observations

* Students who completed the test preparation course achieved higher scores across all subjects.

* The average overall score of students who completed the course was 72.67, compared to 65.04 for students who did not complete it.

* Students who completed the course scored approximately 7–10 points higher on average in each subject.

* Reading and Writing showed particularly strong improvements among students who completed the preparation course.

* The results indicate that test preparation has a positive impact on student academic performance.

In [18]:
# lunch Type Analysis

df.groupby("lunch") [
    ["math score", 
     "reading score", 
     "writing score", 
     "average score"
    ]
].mean()

,math score,reading score,writing score,average score
lunch,,,,
free/reduced,58.921127,64.653521,63.022535,62.199061
standard,70.034109,71.654264,70.823256,70.837209


## Observations

* Students with standard lunch achieved significantly higher scores across all subjects compared to students receiving free/reduced lunch.

* The average overall score of students with standard lunch was 70.84, while students with free/reduced lunch achieved an average score of 62.20.

* The performance gap is approximately 8.6 points in overall average score.

* Students with standard lunch consistently outperformed students with free/reduced lunch in Mathematics, Reading, and Writing.

* The results suggest that lunch type is associated with differences in academic performance.

In [19]:
# Parental Education Analysis

df.groupby("parental level of education") [
    ["math score", 
     "reading score", 
     "writing score", 
     "average score"
    ]
].mean()

,math score,reading score,writing score,average score
parental level of education,,,,
associate's degree,67.882883,70.927928,69.896396,69.569069
bachelor's degree,69.389831,73.000000,73.381356,71.923729
high school,62.137755,64.704082,62.448980,63.096939
master's degree,69.745763,75.372881,75.677966,73.598870
some college,67.128319,69.460177,68.840708,68.476401
some high school,63.497207,66.938547,64.888268,65.108007


## Observations

* Students whose parents hold a master's degree achieved the highest overall average score (73.60).

* Students whose parents hold a bachelor's degree also demonstrated strong academic performance, with an average score of 71.92.

* Students whose parents completed only high school recorded the lowest overall average score (63.10).

* A general upward trend can be observed between parental education level and student academic performance.

* Students with parents holding higher educational qualifications tend to achieve better scores across Mathematics, Reading, and Writing.